File: 088/MergeSortedArray.scala
object MergeSortedArray {
  def merge(nums1: Array[Int], m: Int, nums2: Array[Int], n: Int): Unit = {
    @annotation.tailrec
    def merge(m: Int, n: Int, index: Int): Unit = {
      if (index < 0) Unit
      else if (m >= 0 && n >= 0) {
        if (nums1(m) > nums2(n)) {
          nums1(index) = nums1(m)
          merge(m-1, n, index-1)
        } else {
          nums1(index) = nums2(n)
          merge(m, n-1, index-1)
        }
      } else if (n >= 0) {
        nums1(index) = nums2(n)
        merge(m-1, n-1, index-1)
      }
    }

    merge(m-1, n-1, m+n-1)
//    println(nums1)
  }

  def main(args: Array[String]): Unit = {
    var nums1 = Array(1,2,3,0,0,0)
    var nums2 = Array(2,5,6)
    println(merge(nums1, 3, nums2, 3))
  }
}

File: 088/MergeSortedArray.kt
class Solution {
    fun merge(nums1: IntArray, m: Int, nums2: IntArray, n: Int): Unit {
        var i = m - 1
        var j = n - 1
        var k = n + m - 1
        while (i > -1 && j > -1) nums1[k--] = if (nums1[i] > nums2[j]) nums1[i--] else nums2[j--]
        while (j > -1) nums1[k--] = nums2[j--]
    }
}

fun main(args: Array<String>) {
    val nums1 = intArrayOf(1,2,3,0,0,0)
    val nums2 = intArrayOf(2,5,6)
    val sol = Solution()
    sol.merge(nums1, 3, nums2, 3)
    println(nums1)
}